File: leetcode/C++/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head=l1, *p;
        int c=0;
        for(; l1&&l2; l1=l1->next,l2=l2->next){
            l1->val+=l2->val+c;
            c=l1->val/10;
            l1->val%=10;
            p=l1;
        }
        if(l2) {
            p->next=l2;
            l1=l2;
        }
        for(; l1; l1=l1->next){
            if(c==0) break;
            l1->val+=c;
            c=l1->val/10;
            l1->val%=10;
            p=l1;
        }
        if(c) p->next = new ListNode(c);
        return head;
    }
};

File: leetcode/Python/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        h, p, c = l1, None, 0
        while l1 and l2:
            l1.val+=l2.val+c
            c=l1.val//10
            l1.val=l1.val%10
            p, l1, l2 = l1, l1.next, l2.next
        if l2:
            p.next=l2
            l1=l2
        while l1:
            if c==0:
                break
            l1.val+=c
            c=l1.val//10
            l1.val=l1.val%10
            p, l1 = l1, l1.next
        if c>0:
            p.next=ListNode(val=c)
        return h
        